### PR TITLE
builtin/interfaces: add shared memory interface

### PR DESCRIPTION
**Pull request by mardy**
This interface allows two snaps to communicate to each other over POSIX shared memory object declared in the snap manifest.

The two snaps involved in the connection have exactly the same rights; however, their manifests will be asymmetric:
- the snap providing the slot will be the one which declares the name of the SHM paths to be shared; it will also declare a name for the connection in the "shared-memory" attribute;
- the snap providing the plug will only declare the name of the connection in the "shared-memory" attribute, which will have to match the snap slot's name

### CODE REVIEW

File: interfaces/builtin/shared_memory.go
Comment (zyga) [inline interfaces/builtin/shared_memory.go:33]: Clarity nitpick: this doesn't allow any SHM, it mainly allows snaps to communicate with each other with specific SHM objects. Not sure how to word this properly.

Comment (mardy) [reply interfaces/builtin/shared_memory.go]: I added "predefined" :-)

File: interfaces/builtin/shared_memory.go
Comment (MiguelPires) [inline interfaces/builtin/shared_memory.go:123]: Same here

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:39]: for consistency with the content interface, I think this should also have gadget as a slotting snap type?

File: interfaces/builtin/shared_memory.go
Comment (MiguelPires) [inline interfaces/builtin/shared_memory.go:68]: If `sharedMemory` is asserted to be a string, the second condition can be simplified to `sharedMemory == ""`, right?

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:47]: I think we need to make this super-privileged for now since as I mentioned in the design doc, we don't have a way to know whether a snap is "actually" providing the file that is being granted access to, i.e. a snap may say it is providing access to /dev/shm/org.foo.secret by declaring that in the plug side, but really on the device /dev/shm/org.foo.secret is something that comes from outside snap confinement and the snaps can now access that even though the snaps are not creating the file. This is since /dev/shm is shared with the host filesystem.

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:72]:
```suggestion
		// shared-memory defaults to "slot" name if unspecified
```

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:77]:
```suggestion
		return fmt.Errorf(`shared-memory "write" attribute must be a list of strings, not %v`, value)
```

File: interfaces/builtin/shared_memory.go
Comment (MiguelPires) [inline interfaces/builtin/shared_memory.go:62]:
```suggestion
		BaseDeclarationSlots: sharedMemoryBaseDeclarationSlots,
		AffectsPlugOnRefresh: true,
```

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:110]: is there a reason this uses a Writer instead of just returning the string itself?

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:93]: I would just make this 

```suggestion
			return pathError(pathList)
```

then the message I suggested above is more readable (it says "must be a list of strings not [12, 3, Mixed]" etc which is more clear to me)

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
lgtm for the most part, but I think we need the interface to be super-privileged for now

File: interfaces/builtin/shared_memory.go
Comment (MiguelPires) [inline interfaces/builtin/shared_memory.go:95]: Ah our good ol' friend, the terminating semi-colon :slightly_smiling_face: 
```suggestion
		path = strings.TrimSpace(path)
```

**Review by MiguelPires — APPROVED** [review]
I'm not very deep into this code so just had a couple of minor nitpicks

**Comment by MiguelPires** [comment]
Seems like it's also missing a `gofmt`

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go:110]: well, there's some "cargo programming" here :-), but I do actually prefer it like this, otherwise we'd need to return a concatenated string (or list of strings). Using a writer looks cleaner to me, and more robust if we need to add more rules in the future.

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go:39]: Added!

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: OK. Please check if the changes I made now are correct :-)

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: Done

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: Done

File: interfaces/builtin/shared_memory.go
Comment (stolowski) [inline interfaces/builtin/shared_memory.go:64]: Shouldn't we throw an error if the attribute is not a string?

File: interfaces/builtin/shared_memory.go
Comment (stolowski) [inline interfaces/builtin/shared_memory.go:118]: Same question (about string).

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go:64]: Yes, that would be more correct, on the other hand I copied the code from https://github.com/snapcore/snapd/blob/master/interfaces/builtin/content.go#L74-L81 so I thought this would be fine too :-)

But I'll add the error return.

Comment (mardy) [reply interfaces/builtin/shared_memory.go:64]: This seems to work fine. Let me know if you'd like me to make this change, or if you can suggest some improvement:
```go
    sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]                                             
    sharedMemory, ok := sharedMemoryAttr.(string)
    if isSet && !ok {                                                                                  
        return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,          
            slot.Attrs["shared-memory"])
    }       
    if sharedMemory == "" {                                                                            
```

Comment (stolowski) [reply interfaces/builtin/shared_memory.go:64]: Looks reasonable, one idea to consider but no strong opinion if it's better:
```   
    var sharedMemory string
    sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]                                             
    if isSet {
        var ok bool
        sharedMemory, ok = sharedMemoryAttr.(string) 
        if !ok {                                                                           
            return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,          
               slot.Attrs["shared-memory"])
        }
    }       
    if sharedMemory == "" { 
```                                                                       

We do want to provide a default value right? Because otherwise we could use .Attr(..) getter which does type-checks - however it also complains about missing attribute (but doesn't have distinct error types).

Comment (mardy) [reply interfaces/builtin/shared_memory.go:64]: Yes, the whole point is to gracefully allow for the attribute to be missing. I'll go for my version of the code then, as I find it easier to read.

Comment (mardy) [reply interfaces/builtin/shared_memory.go]: Done

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: And done

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [reply interfaces/builtin/shared_memory.go:111]: using a writer like this is fine, thanks

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:37]: hmm need to think about this, what we typically do for super-privileged snaps is to use `allow-installation: false`, but usually that's on the _plug_ side not on the _slot_ side, but here I think we want the super-privilege settings on the slot side. 

This does have some precedence, we do this for example with the `lxd` and `docker` interfaces, so I think what you want for the base-declaration for the slot side is:

```yaml
  shared-memory:
    allow-installation: false
    deny-connection: true
    deny-auto-connection: true
```

but that does leave a question about what to use for setting up auto-connection rules, since the precedence is that the plug over-rules the slot, so if we put that into the slot, then we could try something like this in the plug declaration:

```yaml
  shared-memory:
    allow-installation: true
    allow-connection:
      plug-attributes:
        shared-memory: $SLOT(shared-memory)
    allow-auto-connection: 
      plug-publisher-id:
        - $SLOT_PUBLISHER_ID
      plug-attributes:
        content: $SLOT(content)
```

but importantly because the plug rules will over-ride the slot rules this means that a snap with a plug would immediately over-ride the super-privileged nature of the slot and it would connect. However, I think it still makes sense because the `allow-installation: false` on the slot means that the snap with the slot can not even be installed without a store snap-declaration assertion overriding the `allow-installation` setting to `true`, so maybe this all makes sense.

In any case I would like to hear from @pedronis to confirm my understanding is correct and that the two snippets above are what we should put for the slot and plug side respectively.

**Comment by anonymouse64** [comment]
marking blocked until we can get clarity on the desired base-declaration for this interface, hopefully what I suggested is correct though and this can be unblocked with a simple "yes that's correct' :smile:

File: interfaces/builtin/shared_memory.go
Comment (pedronis) [reply interfaces/builtin/shared_memory.go:37]: yes, those rules would make a snap with a plug follow rules like the content interface,  notice that I think you didn't fully update the snippet, we want I think:

```
allow-auto-connection: 
      plug-publisher-id:
        - $SLOT_PUBLISHER_ID
      plug-attributes:
        share-memory: $SLOT(shared-memory)
```

slot snaps cannot be installed by default, as they are the ones that can select the shared file, it seems ok

**Review by pedronis — COMMENTED** [review]
this overall looks good to me apart the base-declaration considerations, also it will need a security review

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:81]: so reading the spec and thinking about this more, I'm thinking about how we can support the read/write aspect of shared-memory to match content and system-files. 

For background, with the content interface, snaps can setup one way communication where the directory shared is writable from the snap providing the slot just by choosing a directory like $SNAP_DATA as the specifier in `read`, and then the plug side of connection only has rules allowing read access to the files shared. The slot can setup bi-directional sharing by choosing a directory that the slotting snap can write to as the source in the `write` setting, and then the plug side is allowed to write to that directory. 

However, we can't do the same thing here for shared-memory, since if we used the same pattern and had something like this:

```
slots:
  foo-mem:
    interface: shared-memory
    read:
      - /dev/shm/org.foo.mem
```

Then we would setup read-only permissions for the plug, but then what permissions do we setup for the slot side? We should arguably always setup write permissions to enable the one way communication from slot -> plug.

So given that, I think that what we should do is use the slot side settings for read/write to indicate the permissions for the plug side, but always specify write access for the slot.

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
I think my spec was not sufficiently thought through regarding the behavior of the read/write settings, but I think what we should do is have the slot side setting of read/write affect the permission of the plug side, but the slot side always gets write permission to the /dev/shm object. I will update my spec to reflect this

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:119]: also this does not permit the slot to specify both the short hand and the long hand, i.e. both

```yaml
slots:
  foo-mem:
    interface: shared-memory
    write: [ /dev/shm/org.foo.mem ]
```

and 

```yaml
slots:
  foo-mem:
    interface: shared-memory
    write: [ org.foo.mem ]
```

Comment (mardy) [reply interfaces/builtin/shared_memory.go:119]: Do we already know that we'll want to support paths other than those used by `shm_open()`? If not, I'd rather leave this to a follow-up PR, to be made when the use-case is clear.

Comment (anonymouse64) [reply interfaces/builtin/shared_memory.go]: The point here is not about supporting paths outside of `/dev/shm/`, the point is about accepting both forms of the interface definition from the spec (both of my examples above should result in the same policy written out to disk). Ideally we would validate that the long-form specification only has `/dev/shm` as the prefix in the long-hand. You could switch on if the first character is `/` or not. 

I personally don't think this should be a huge change to support both the absolute and the relative paths, but I do think if you only want to support the short-hand in this PR and add support for the long-hand in another PR that there should be an error for the absolute path case which AFAICT there is not today

**Comment by atesburak** [comment]
Just asking out of curiosity, could a single snap declare the same SHM with multiple other snaps?

**Comment by mardy** [comment]
> Just asking out of curiosity, could a single snap declare the same SHM with multiple other snaps?

Yes. Normally you would have a single snap declaring the SHM in its slot, and one or more snaps declaring it in their plugs. 

It's indeed also possible that different snaps declare the same SHM object in their slot (possibly because they are developed by different people), and in that case they would actually be referring at the very same SMH area at runtime. WHich might or might now cause issues. In practice, this is a privileged interface so I'd expect the store team to review the accesses and raise a flag if the same SMH object has been already declared by another snap.

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: I see. Well, since there is a small window for further discussions (like, should we also allow `/run/shm/...`?), I'll just take the second option and add an error code if the path contains a `/`.

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: Done!

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:61]: should we also do a check using `cleanSubPath` right? I suppose it's unnecessary for now since we disallow `/` entirely, but maybe when we support the long-form we should add that check too

File: interfaces/builtin/shared_memory.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory.go:178]: thanks for this

File: interfaces/builtin/shared_memory_test.go
Comment (anonymouse64) [inline interfaces/builtin/shared_memory_test.go:209]: could you add two more test cases, with both read and write where one is valid and the other is invalid?

File: tests/main/interfaces-shared-memory/shm-slot/meta/snap.yaml
Comment (anonymouse64) [inline tests/main/interfaces-shared-memory/shm-slot/meta/snap.yaml:11]: could you either add a path that includes a regular expression or modify this one to be a pattern so we can show that the regular expression works too?

**Review by anonymouse64 — APPROVED** [review]
lgtm, a few little nitpicks about tests but otherwise is looking really good and I think is quite close

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go:61]: Actually I realized that I was still allowing ".." as path; now I added `cleanSubPath`, and that fixed it. Thanks for pointing that out :-)

File: interfaces/builtin/shared_memory_test.go
Comment (mardy) [reply interfaces/builtin/shared_memory_test.go:213]: Added.

File: tests/main/interfaces-shared-memory/shm-slot/meta/snap.yaml
Comment (mardy) [reply tests/main/interfaces-shared-memory/shm-slot/meta/snap.yaml:11]: Uh! This fails, because in the validation code I have a call to `apparmor.ValidateNoAppArmorRegexp()` -- I obviously must not have noticed that part of the spec where you talk about supporting "`*`". Since the handling of it is not trivial (I added a question to the spec document, please have a look), I wonder if we can merge this without it? I added a TODO comment.

Comment (anonymouse64) [reply tests/main/interfaces-shared-memory/shm-slot/meta/snap.yaml:11]: Ah I see, sorry I keep forgetting that all your code about validating AARE is still not merged yet :-) 

I think it's fine in this case to not support regular expressions in this first iteration in this PR, and then after mount-control step 2 PR lands with the AARE stuff then I think we can tie that validation into this, but to be clear we do definitely need to allow wild-cards specifically in the final version of the interface as that's critical for customer use-cases.

**Comment by alexmurray** [comment]
> In practice, this is a privileged interface so I'd expect the store team to review the accesses and raise a flag if the same SMH object has been already declared by another snap.

There are multiple people who are reviewers for the store and hence who grant store declarations. There is no central repository of all declarations that can easily be queried so I don't think this is feasible to expect reviewers to not grant multiple shm declarations for the same object to different snaps. Perhaps we could have a policy where snaps should use a shm object name which is somehow related to the snap name / reverse DNS style name of the app itself - then we would be able to spot instances where a snap is trying to access a shm object which it doesn't own (similarly to the current review process for DBus slot names).

**Review by alexmurray — APPROVED** [review]
LGTM, thanks for the comprehensive tests - when the additional PR for adding wildcards to this is proposed please mark this as needs-security-review too.

**Comment by anonymouse64** [comment]
@alexmurray I agree that a policy for the global store should be developed, but I also think that such a tool for viewing all such granted interfaces should be developed as this would help greatly in evaluating requests. Additionally, I think the comment was more geared towards private/brand-store only snaps where the queue is smaller and more lenient than the global store.

File: interfaces/builtin/shared_memory.go
Comment (pedronis) [inline interfaces/builtin/shared_memory.go:83]: nitpick: read is unusual here as a prefix,  it seems we tend to use nothing or "get" or "verify". Also here it might seem to refer to the "read" attribute but is actually for both read and write. My preference is to just call this:  "stringListrAttribute".

**Review by pedronis — APPROVED** [review]
I looked only at the interface code itself, looks good, one nitpick.

File: interfaces/builtin/shared_memory.go
Comment (mardy) [reply interfaces/builtin/shared_memory.go]: Done!

**Pull request merged**
